New class JExtensionHelper providing a safe way to determine if an extension is core or not

richard67 opened the pull request:
Pull Request for Issue #16579 .

### Summary of Changes

New class JExtensionHelper for helping with core extensions detection.

Thanks to @andrepereiradasilva who did the first implementation. I extended that a bit.

As soon as this is finished and merged, I will make another PR if necessary to replaces existing tests in SQL statements for extension_id > 999 or similar by using 'NOT (' . JExtensionHelper::getWhereCondition() . ')' as a query condition or using the JExtensionHelper::getCoreExtensionsIds() or JExtensionHelper::getCoreExtensionsIdsList() functions if that would be more appropriate.

And I will make similar PRs to the repositories of the authors of PRs #16474 and #16494 if those PRs are still open then, otherwise they will have been caught with the previous step.

### Testing Instructions

1. Code reviews by experienced manitainers are welcome
2. Test the new class as follows:
- Have error reporting set to maximum in global configuration, and if possible let php log errors into a file by setting log_errors to On and error_log to the full path to a log file in your php.ini.
- Add the following code to your frontend template's index,php, e.g. in case of the protostar template just after `<div class="body" id="top">`:
```
<p style="font-weight: bold;"><?php echo "JExtensionHelper::getWhereConditionCore():"; ?><p>
<p><?php $test1=JExtensionHelper::getWhereConditionCore(); var_dump($test1); ?><p>
<p style="font-weight: bold;"><?php echo "JExtensionHelper::getWhereConditionNonCore():"; ?><p>
<p><?php $test1=JExtensionHelper::getWhereConditionNonCore(); var_dump($test1); ?><p>
<p style="font-weight: bold;"><?php echo "JExtensionHelper::getCoreExtensions():"; ?><p>
<p><?php $test1=JExtensionHelper::getCoreExtensions(); var_dump($test1); ?><p>
<p style="font-weight: bold;"><?php echo "JExtensionHelper::getCoreExtensionsIds():"; ?><p>
<p><?php $test1=JExtensionHelper::getCoreExtensionsIds(); var_dump($test1); ?><p>
<p style="font-weight: bold;"><?php echo "JExtensionHelper::getNonCoreExtensionsIds():"; ?><p>
<p><?php $test1=JExtensionHelper::getNonCoreExtensionsIds(); var_dump($test1); ?><p>
<p style="font-weight: bold;"><?php echo "JExtensionHelper::getCoreExtensionsIdsList():"; ?><p>
<p><?php $test1=JExtensionHelper::getCoreExtensionsIdsList(); var_dump($test1); ?><p>
<p style="font-weight: bold;"><?php echo "JExtensionHelper::getNonCoreExtensionsIdsList():"; ?><p>
<p><?php $test1=JExtensionHelper::getNonCoreExtensionsIdsList(); var_dump($test1); ?><p>
<p style="font-weight: bold;"><?php echo "JExtensionHelper::checkIfCoreExtension('component', 'com_mailto', '', 0):"; ?><p>
<p><?php $test1=JExtensionHelper::checkIfCoreExtension('component', 'com_mailto', '', 0); var_dump($test1); ?><p>
<p style="font-weight: bold;"><?php echo "JExtensionHelper::checkIfCoreExtension('component', 'com_blabla', '', 0):"; ?><p>
<p><?php $test1=JExtensionHelper::checkIfCoreExtension('component', 'com_blabla', '', 0); var_dump($test1); ?><p>
<p style="font-weight: bold;"><?php echo "JExtensionHelper::checkIfCoreExtensionId(1):"; ?><p>
<p><?php $test1=JExtensionHelper::checkIfCoreExtensionId(1); var_dump($test1); ?><p>
<p style="font-weight: bold;"><?php echo "JExtensionHelper::checkIfCoreExtensionId(10000):"; ?><p>
<p><?php $test1=JExtensionHelper::checkIfCoreExtensionId(10000); var_dump($test1); ?><p>
```

### Expected result

The result should look similar to follows (long stuff replaced by "..."):

```
JExtensionHelper::getWhereConditionCore():

string(13256) "`type` = 'component' AND `element` = 'com_admin' AND `client_id` = 1 OR `type` = 'component' AND `element` = 'com_ajax' AND `client_id` = 1 OR ... OR 1 = 2"

JExtensionHelper::getWhereConditionNonCore():

string(13900) "(`type` <> 'component' OR `element` <> 'com_admin' OR `client_id` <> 1) AND (`type` <> 'component' OR `element` <> 'com_ajax' OR `client_id` <> 1) AND ...  AND 1 = 1"

JExtensionHelper::getCoreExtensions():

array(161) { [0]=> array(4) { [0]=> string(9) "component" [1]=> string(10) "com_mailto" [2]=> string(0) "" [3]=> int(0) } [1]=> array(4) { [0]=> string(9) "component" [1]=> string(11) "com_wrapper" [2]=> string(0) "" [3]=> int(0) } ... }

JExtensionHelper::getCoreExtensionsIds():

array(161) { [0]=> string(1) "1" [1]=> string(1) "2" ... }

JExtensionHelper::getNonCoreExtensionsIds():

array(15) { [0]=> string(5) "10000" [1]=> string(5) "10001" ... }

JExtensionHelper::getCoreExtensionsIdsList():

string(608) "1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,22,23,24,25,27,28,29,30,31,32,33,34,102,103,104, ... ,700,802,10041,10042,10043"

JExtensionHelper::getNonCoreExtensionsIdsList():

string(89) "10000,10001,10002,10004,10005,10006,10044,10045,10055,10060,10062,10079,10086,10109,10110"

JExtensionHelper::checkIfCoreExtension('component', 'com_mailto', '', 0):

bool(true)

JExtensionHelper::checkIfCoreExtension('component', 'com_blabla', '', 0):

bool(false)

JExtensionHelper::checkIfCoreExtensionId(1):

bool(true)

JExtensionHelper::checkIfCoreExtensionId(10000):

bool(false) 

```

Verify that the comma-separated list of extension_ids which has been produced by JExtensionHelper::getCoreExtensionsIdsList() contains only the IDs of core extensions, especially if there are IDs greater than or equal to 10000. Also those should belong to core extensions, the list should not contain any IDs of non-core extensions. For the others (smaller than 10000) it is enough to spot check some of them, checking all it too much.

Verify that the comma-separated list of extension_ids which has been produced by JExtensionHelper::get**Non**CoreExtensionsIdsList() contains only the IDs of **non-**core extensions.

Check that no PHP errors are shown or logged in your log file (if you use some).

### Actual result

There is no actual result because it is a new class.

### Documentation Changes Required

None.

richard67 committed:
New class JExtensionHelper

New class JExtensionHelper for helping with core extensions detection

richard67 committed:
Code style, change init function to protected

richard67 committed:
Make drone happy

richard67 committed:
Make initCoreExtensionsIds public again

Make function initCoreExtensionsIds() public again so it can be used to
rebuild the array e.g. after a discovery installation.

richard67 committed:
Get the extension_id ordered ascending

Get the extension_id array ordered ascending - this will speed up
queries using "IN (...)" or "NOT IN (...)" as condition using function
getCoreExtensionsIdsList(), and it should also speed up the in_array()
check in function checkIfCoreExtensionID().

richard67 committed:
Correct last commit - add missing change

richard67 committed:
Raise an exception if no data found

Raise an exception if no core extensions have been found in database
when getting the list of installed extension_ids

Quy commented on libraries/cms/extension/helper.php:
```diff
@@ -0,0 +1,381 @@
+	}
+
+	/**
+	 * Gets the core extensions ids as comma-separated list which can be used
```
Change `ids` to `IDs`

richard67 replied:
yep, will do soon. am still working on it.

Quy commented on libraries/cms/extension/helper.php:
```diff
@@ -0,0 +1,381 @@
+	 * to get only core extensions or "WHERE extension_id NOT IN (1,2,3)" to 
+	 * exclude core extensions.
+	 *
+	 * @return  string  Comma-separated list of core extension ids.
```
Change `ids` to `IDs`

richard67 replied:
yep, will do soon. am still working on it.

Quy commented on libraries/cms/extension/helper.php:
```diff
@@ -0,0 +1,381 @@
+			self::initCoreExtensionsIds();
+		}
+
+		return implode(",", self::$coreExtensionsIDs);
```
Use single quote `','`

richard67 replied:
OK, will be commited soon.

Quy commented on libraries/cms/extension/helper.php:
```diff
@@ -0,0 +1,381 @@
+	 * @param   string   $type       The extension's type.
+	 * @param   string   $element    The extension's element name.
+	 * @param   string   $folder     The extension's folder.
+	 * @param   integer  $client_id  The extension's client_id.
```
Change `client_id` to `client ID`

richard67 replied:
OK, will be commited soon.

richard67 committed:
Code style + spelling + optimization

Code style + spelling + build the static where clause for the query only
once on class construction + new function to get that where clause +
extended descriptions in doc blocks

richard67 committed:
Optimize script.php

Let script.php not do the same work again as the new helper class does
with building the where clause for the database query.

richard67 committed:
Fixed errors from previous commit

richard67 committed:
Correct function comment for drone

richard67 commented:
@Quy I have implemented your review comments. As soon as AppVeyor has finished with success, all is ready I hope.

Questions:
1. Will it need unit tests for the new helper class? If so, what all should they test?
2. Is there any developer documentation about how to add extensions to Joomla core?
If so, it would have to be extended by the information that the array in the new helper class has to be updated, too, if adding a new extension.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16613">issues.joomla.org/tracker/joomla-cms/16613</a>.</sub>

zero-24 commented:
@richard67 can you please add doc block comments for the propertyes too? Basicly Usage, type (using @var ) and since tag 😄

richard67 committed:
Add doc block comments also for variables

Add doc block comments also for variables and not only for functions.

richard67 commented:
@zero-24 Done. Thanks for the hint, also thanks to @Quy for the review before.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16613">issues.joomla.org/tracker/joomla-cms/16613</a>.</sub>

richard67 committed:
Improved a dock block comment

richard67 committed:
Again a correction for doc block comment.

Quy commented on libraries/cms/extension/helper.php:
```diff
@@ -0,0 +1,456 @@
+
+		// Core module extensions.
+		// - Site
+		array('module', 'mod_articles_archive', '', 0),
```
Sort in alpha order to be consistent with the other sections???

richard67 replied:
Well other sections party are not sorted too. The order is more or less like it came with joomla.sql in the insert. But it could make sense to order them all alphabetically for better maintainability. Should I do that?

richard67 replied:
Or should new extensions of a certain type just be appended to the end of the corresponding section if someone adds extensions in future?

Quy replied:
In myPHPAdmin, i searched for module and the results were in alphabetical order. I did not search for the other types to confirm. For maintainability, I would say yes, but it is really up to you if you don't mind and have the time.

Quy replied:
Yes to separate from the core and non-core extensions.

richard67 replied:
How do you mean your last comment?

Quy replied:
I think you mean for new core extensions. If yes, it should still be in alpha order just like language keys/strings.

Quy replied:
Ignore my comment before the last comment :smile

richard67 committed:
Alpha-order the array elements in each section

andrepereiradasilva commented:
@richard67  i must don't agree in putting core extensions stuff in the constructor. I mean that class can be used in the future for more method that are not core extensions related. It's called jextensionhelper, not jcoreextensionshelper ;)

richard67 commented:
@andrepereiradasilva  Which constructor? I changed it back to a normal callable function a few commits before.

Quy commented on libraries/cms/extension/helper.php:
```diff
@@ -0,0 +1,457 @@
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public static function checkIfCoreExtensionID($extension_id)
```
Change `ID` to `Id` for consistency.

andrepereiradasilva commented:
Oh forget it. Commenting on old version. Ignore.

richard67 committed:
Renamed core-related things, synch spelling of ID

Renamed core-related things and extended block comments so it is clear
what is related to core extensions only, synch spelling of ID in
function and varibale names to Id so it is only ID in comments.

richard67 commented:
@Quy Done the ID to id thing, also at a few more places.

@andrepereiradasilva Renamed things which are related to core extensions only so it is clear and more open for future extensions by handling non-core extensions, also extended doc block comments.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16613">issues.joomla.org/tracker/joomla-cms/16613</a>.</sub>

richard67 commented:
@andrepereiradasilva By the way, I checked against joomla.sql of current staging and the array was complete, nothing was missing or wrong. Thanks for your work for this.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16613">issues.joomla.org/tracker/joomla-cms/16613</a>.</sub>

Quy commented on administrator/components/com_admin/script.php:
```diff
@@ -380,199 +380,14 @@ protected function removeJedUpdateserver()
-		);
-
 		// Attempt to refresh manifest caches
+		$whereCondition = JExtensionHelper::getWhereCondition();
```
Update name.

richard67 replied:
Done.

Quy commented on libraries/cms/extension/helper.php:
```diff
@@ -0,0 +1,458 @@
+	{
+		if (self::$whereConditionCore === '')
+		{
+			self::initWhereCondition();
```
Update name.

richard67 replied:
Done.

richard67 committed:
Correct errors in previous commit

Thanks to @Quy rename things forgotten in previous commit.

richard67 commented:
@Quy Thanks for review.

Quy commented:
Update testing instructions: `<p><?php $test1=JExtensionHelper::getWhereCondition(); var_dump($test1); ?><p>`

Quy commented:
Should be `less than`:
`Check that if the comma-separated list of extension_ids which has been produced by JExtensionHelper::getCoreExtensionsIdsList() contains values greater than or equal to 10000, then these extensions belong to the core.`

richard67 commented:
@Quy Testing instructions alrerady had been updated.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16613">issues.joomla.org/tracker/joomla-cms/16613</a>.</sub>

richard67 commented:
@Quy No, i mean those which are in the list and greater or equal than 1000. For those it has to be checked if they are core, otherwise they should not be in the list, and for those smaller than 10000 it is not really necessary to check because they are very likely core, but of course someone can check this, too.

Quy commented:
Do you mean this ? `then these extensions do not belong to the core.`

richard67 commented:
@Quy No. I updated testing instructions to make it clear.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16613">issues.joomla.org/tracker/joomla-cms/16613</a>.</sub>

richard67 commented:
Installations which have gone through updates which have addded new core extensions during their history will have core extensions with IDs greater than or equal to 10000.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16613">issues.joomla.org/tracker/joomla-cms/16613</a>.</sub>

Quy commented:
I have tested this item :white_check_mark: successfully on a186ccf3766badaa02394707c99c7b1032f7f82f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16613">issues.joomla.org/tracker/joomla-cms/16613</a>.</sub>

laoneo commented:
I don't see the problem you are trying to solve. But to determine if an extension belongs to core I would add a new property (column) or even introduce a new state for an extension. For me it is hard to understand that we must point devs to that file when they want to add a new extension to the core.

brianteeman commented:
1. It is rare that we add or remove an extension 
2. It is no harder than adding a removed file to script.php

Bakual commented:
@laoneo As you can see in the PR, we already maintain such a list so the manifest of core can be updated during a CMS update. Just that this list currently isn't available for other use cases. This PR creates an helper method which can be used when needed.
A new column in the database was discussed, but to me doesn't make sense since this is static information which is always the same for a given Joomla version. Imho it doesn't make sense to store such information into the database.

richard67 committed:
Add functions for non-core extensions

Add a static SQL where clause for non-core extensions based on optimized
negated where clause for core extensions, add functions to get non-core
extensions IDs.

richard67 commented:
@Quy Sorry but I thought it is useful to add some new functions for non-core extensions. Could you test and review again? That would be fine.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16613">issues.joomla.org/tracker/joomla-cms/16613</a>.</sub>

Quy commented on libraries/cms/extension/helper.php:
```diff
@@ -0,0 +1,611 @@
+	 * Note that this causes an extra database query to the extensions table
+	 * to get the IDs, so it is only economic if you need this array of IDs later
+	 * in the code. Using function getWhereConditionNonCore() to get a where clause to
+	 * restrict your queries to core extensions will not cause such an extra read
```
Change to `non-core`

richard67 replied:
Right. Will correct in some 1 hour.

richard67 replied:
Done.

Quy commented on libraries/cms/extension/helper.php:
```diff
@@ -0,0 +1,611 @@
+	 * Note that this causes an extra database query to the extensions table
+	 * to get the IDs, so it is only economic if you need this list of IDs later
+	 * in the code or more than 1 time. Using function getWhereConditionNonCore() to
+	 * get a where clause to restrict your queries to core extensions will not
```
Change to `non-core`

richard67 replied:
Right. Will correct in some 1 hour.

richard67 replied:
Done.

Quy commented on libraries/cms/extension/helper.php:
```diff
@@ -0,0 +1,611 @@
+
+		return self::$whereConditionCore;
+	}
+	/**
```
Add a blank line above.

richard67 replied:
Will do as soon as this slow Windows 10 creator's update has finished here so I am online again with my PC. Thanks so far for your reviews, I really appreciate them.

richard67 replied:
But no need to tag this PR with 5.0 now ... I don't think it will be this slow.

richard67 replied:
Done.

richard67 committed:
Code style and doc block comment corrections

Code style and doc block comment corrections as found by @Quy during
review.

Quy commented on libraries/cms/extension/helper.php:
```diff
@@ -0,0 +1,612 @@
+		{
+			self::$whereConditionCore .= $db->qn('type') . ' = ' . $db->q($extension[0])
+				. ' AND ' . $db->qn('element') . ' = ' . $db->q($extension[1])
+				. ' AND ' . $db->qn('client_id') . ' = ' . $db->q($extension[3]);
```
`client_id` is a tinyint so don't need to be enclosed in quotes.

richard67 replied:
Done.

richard67 committed:
Don't quote integer values

Don't quote integer values in where clauses

Quy commented:
I have tested this item :white_check_mark: successfully on 93b4fd64babb2071f0339bc013be04d17657136e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16613">issues.joomla.org/tracker/joomla-cms/16613</a>.</sub>

alikon commented:
i really like the concept but 
![nopleaseno](https://user-images.githubusercontent.com/181681/27195819-82501fa2-5208-11e7-80f8-8b9ca375a4d4.PNG)
please not in this way....
queries are very ugly to me :alien:

richard67 commented:
Which else way? The query is the same as it was in script.php before this PR. It is nothing new, only has been moved into the new class.

alikon commented:
ok
 i was stuck on  pr title 

> New class JExtensionHelper

richard67 commented:
@alikon Do you think some index could speed up the queries?
Maybe
ALTER TABLE `j3ux0_extensions` ADD KEY `idx_type_element_folder_client_id` (`type`, `element`, `folder`, `client_id`);
or
ALTER TABLE `j3ux0_extensions` ADD KEY `idx_type_element_client_id` (`type`, `element`, `client_id`);
or indexes on the individual columns?

alikon commented:
i don't think the main problem here is about perfomance, rather, is much more on the data model design side,
in my opinion  adding a boolean field `'core_extenions'`  could be a good choice....

richard67 commented:
@alikon That was discussed before in issue #16579 . The weak point of using a database column for a core or non-core flag is that a silly extensions developer may set it for his/her extensions, too, and so mess up everything.

alikon commented:
forgive me missed that one #16579 :flushed:

laoneo commented:
Are really all this functions needed? Why not just providing a function which returns the extensions. Additionally we can think about a flag, probably something like `public static function getExtensions($onlyCore = false)`? Keep in mind this class should also get some unit tests at the end.

richard67 commented:
@laoneo What do you mean with "which returns the extensions"? One which returns the static array like getCoreExtensions is in my PR right now? In this case a "$onlyCore = false" would not make sense, since we do not have and never wanna have a static array including the non-core extensions. Or one which reads stuff from DB? Please make more clear what you mean, and maybe paste here a doc bloc of some function you have in mind so we can understand better what you mean.

richard67 commented:
The more I think about it, the less I think that if data is static or not is not a criteria whether it should e in database or not, like @Bakual commented a bit above. The columns `type`, `element`, `folder` and `client_id` we use now in script.php and possibly in future with this PR to identify the core extensions are static, too. It makes sense to store stuff in database if it helps the application with having more simple code and/or be much more performant.

A column "is_core" in the extensions table having an index on it whould simplify code much and speed up queries.

To be able to rely on the value in that column, it is important that the schema update settting the values for this columns on update of old installations is run on the update. In case this is not granted, we could still let script.php go through the extensions table like it does now for updating the manifest cache, using the static array with the `type`, `element`, `folder` and `client_id` values.

If we give the new column a default value of 0 (false) we could even not mention it in docs and examples for non-core extension developers, but maybe we should mention it and tell that if they set this to 1 in their extensions, stuff might not work, so they have to set it to 0.

I think about making a new PR with the database column solution as an alternative to this one here.

Bakual commented:
> To be able to rely on the value in that column, it is important that the schema update settting the values for this columns on update of old installations is run on the update. 

Friendly reminder: The reason we discuss this PR in the first place is because we can't rely on the updates being run properly. We can rely on the structure, but not on the data. That's why core extensions *could* have IDs higher than 10000 even if they shouldn't.

Honestly, this PR evolved for some reason beyond the initial idea. Personally I would just store that array with the needed data. Imho, there is no reason for a method doing the SQL query to fetch the IDs. Can't we keep the database stuff out of this class? The class certainly looks very complex now for such a trivial task.

The cases where that list of core extensions is used are very limited. Do we need the IDs in all those cases?

richard67 commented:
Well the idea behind the additional functions beside getCoreExtensions was to avoid duplicate code.
But limiting the new helper to the static aray only and keeping all database stuff out has also its charme.

@laoneo Did you mean same as @Bakual , to limit the new helper to the static array thing only?

richard67 commented:
@laoneo @Bakual I think we should at least keep the checkIfCoreExtension function, too, not only the getCoreExtensions. What do you think about this?

Bakual commented:
>  I think we should at least keep the checkIfCoreExtension function, too, not only the getCoreExtensions. What do you think about this?

Yeah, that one looks nice and helpful.

richard67 commented:
Closing in favour of PR #16724 . Thanks for the reviews and discussions up to now. Further discussions please continue there.
